src/modules/bulk/components/BulkAdd.tsx

```diff
@@ -0,0 +1,250 @@
+import { TypedDocumentNode, useMutation } from '@apollo/client';
+import { Alert, CircularProgress, Paper } from '@mui/material';
+import { Stack } from '@mui/system';
+import { isEmpty } from 'lodash-es';
+import { ReactNode, useMemo, useState } from 'react';
+
+import DynamicField, {
+  DynamicFieldProps,
+} from '@/modules/form/components/DynamicField';
+import useDynamicFormFields from '@/modules/form/hooks/useDynamicFormFields';
+import {
+  buildCommonInputProps,
+  CONFIRM_ERROR_TYPE,
+  FormValues,
+  LocalConstants,
+} from '@/modules/form/util/formUtil';
+import { transformSubmitValues } from '@/modules/form/util/recordFormUtil';
+import {
+  FormDefinitionJson,
+  FormItem,
+  useGetFormDefinitionByIdentifierQuery,
+  ValidationError,
+  ServiceDetailType,
+} from '@/types/gqlTypes';
+
+const extractClientItemsFromDefinition = (definition: FormDefinitionJson) => {
+  const targetItems: FormItem[] = [];
+
+  const traverse = (items: FormItem[]): FormItem[] => {
+    return items.reduce((acc, item) => {
+      if (item.serviceDetailType === ServiceDetailType.Client) {
+        targetItems.push(item);
+        return acc;
+      }
+      if (item.item) {
+        return [
+          ...acc,
+          {
+            ...item,
+            item: traverse(item.item),
+          },
+        ];
+      }
+      return [...acc, item];
+    }, [] as FormItem[]);
+  };
```

Comment:
Let's move this function `extractClientItemsFromDefinition` into `formUtil.ts`.

I also suggest replacing the recursion with a call to `getItemMap` which flattens the form definition. I try to use that wherever possible to avoid repeating the recursion logic all over the place.


```suggestion
const extractClientItemsFromDefinition = (definition: FormDefinitionJson) => {
    const itemMap = getItemMap(definition, false); // flattened map { linkId => item }
    const targetItems = Object.values(itemMap).filter(
      ({ serviceDetailType }) => serviceDetailType === ServiceDetailType.Client
    );
```